File: Data Structures & Algorithms/palindromic-substrings/submission-0.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        res = 0
        seen = set()
        def check(l, r):
        
            curr = 0
            while l >= 0 and r < n and s[l] == s[r]:
                if (l, r) not in seen:
                    curr += 1
                seen.add((l, r))
                l -= 1
                r += 1

            return curr
 
        for l in range(n):
            r = l + 1
            res += check(l, l)
            res += check(l, r)
            
        return res              

File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {        
        vector<int> tmp;
        dfs(nums, 0, 0, tmp, target);
        return res;

    }
    void dfs(vector<int>nums, int i, int curr, vector<int>& tmp, int target) {
            if (curr == target) {
                res.push_back(tmp);
                return; 
            }
            if (i >= nums.size() || curr > target) {
                return;
            }

            tmp.push_back(nums[i]);
            dfs(nums, i, curr + nums[i], tmp, target);
            tmp.pop_back();
            dfs(nums, i + 1, curr, tmp, target);
        }
};
